### PR TITLE
Fix attributes diff on DOM node

### PR DESCRIPTION
**Pull request by dmail**
### Expected behaviour 

Diff should remove attributes when they are no longer present in the next html state (see [unit test](https://github.com/developit/preact/compare/master...dmail-fork:master?expand=1#diff-a1195645838de4f080fcb358b5132a4cR499))

### Actual behaviour

It behaves as expected when the merged node is created using `h`.
It preserves removed attributes when the merged node comes from the DOM.

### Possible solutions

- I suggest to approve this pull request :)

### CODE REVIEW

File: src/vdom/diff.js [hunk 1]
```diff
@@ -124,6 +124,8 @@ function idiff(dom, vnode, context, mountAll, componentRoot) {
 		props = out[ATTR_KEY] || (out[ATTR_KEY] = {}),
 		vchildren = vnode.children;
 
+	for (let a=out.attributes, i=a.length; i--; ) props[a[i].name] = a[i].value;
```
Comment (developit) [inline src/vdom/diff.js]: we'll need to wrap this in a check for props. If the propcache already exists on an element, skipping attribute checks is a fairly major performance gain.

Comment (developit) [inline src/vdom/diff.js]: I'm fine doing this post-merge if you want.

**Comment by dmail** [comment]
@developit code updated to loop over attributes only when not cached : https://github.com/developit/preact/pull/677/files#diff-a0896fa5d65b1d80ca4d496badbae50fR127

File: src/vdom/diff.js [hunk 2]
```diff
@@ -123,9 +123,17 @@ function idiff(dom, vnode, context, mountAll, componentRoot) {
...
 	let fc = out.firstChild,
-		props = out[ATTR_KEY] || (out[ATTR_KEY] = {}),
+		props,
 		vchildren = vnode.children;
 
+	if (out[ATTR_KEY]) {
+		props = out[ATTR_KEY];
+	}
+	else {
+		props = out[ATTR_KEY] = {};
```
Comment (developit) [inline src/vdom/diff.js]: I'll probably consolidate this post-merge, but it's good!

**Comment by dmail** [comment]
Hi @developit, is there a planned release with this fix ?

**Comment by developit** [comment]
Yup, 8.2 out soon.

**Comment by timmywil** [comment]
@developit would love that release!

**Comment by developit** [comment]
Me too, heh.

**Pull request merged**